morphia/src/main/java/org/mongodb/morphia/MorphiaReference.java

```diff
@@ -0,0 +1,123 @@
+/*
+ * Copyright (c) 2008-2016 MongoDB, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.mongodb.morphia;
+
+import com.mongodb.DBCollection;
+import org.mongodb.morphia.query.QueryException;
+import org.mongodb.morphia.query.ValidationException;
+
+public class MorphiaReference<T> {
+    private Object id;
+    private String type;
+    private String collection;
+
+    private transient T entity;
+    private transient Class<?> typeClass;
+
+    private MorphiaReference() {
+    }
+
+    static <T> MorphiaReference<T> toEntity(final T entity, final Datastore datastore) {
+        MorphiaReference<T> ref = new MorphiaReference<T>();
+        ref.entity = entity;
+        ref.typeClass = entity.getClass();
+        ref.type = entity.getClass().getName();
+        Object id = datastore.getKey(entity).getId();
+        ref.id = id;
+        if (id == null) {
+            throw new ValidationException("The referenced entity has no ID.  Please save the entity first.");
+        }
+
+        return ref;
+    }
+
+    static <T> MorphiaReference<T> toEntity(final T entity, final DBCollection collection, final Datastore datastore) {
+        MorphiaReference<T> ref = toEntity(entity, datastore);
+        ref.collection = collection.getName();
+
+        return ref;
+    }
+
+    @SuppressWarnings("unchecked")
+    T fetch(final Datastore datastore) {
+        if (entity == null) {
+            if (collection != null) {
+                entity = (T) ((DatastoreImpl) datastore).getCollection(collection).findOne(id);
```

Comment:
This looks like a call to DBCollection.findOne, which is going to return a DBObject, not a T